File: benchmarks/benchmark_comparison.py
#!/usr/bin/env python
"""Benchmark comparison between Python and OCaml implementations."""

import time
from game import Game2048
from expectimax_python import max_value, expect_value, get_best_move

def simple_eval(board):
    """Simple evaluation function for benchmarking."""
    empty_count = sum(1 for i in range(16) if ((board >> (4*i)) & 0xF) == 0)
    max_tile = 0
    for i in range(16):
        cell = (board >> (4*i)) & 0xF
        if cell > max_tile:
            max_tile = cell
    return float(empty_count) + float(max_tile)

def load_tape(filename):
    """Load random tape."""
    events = []
    with open(filename, 'r') as f:
        for line in f:
            line = line.strip()
            if line and not line.startswith('#'):
                parts = line.split()
                if len(parts) >= 2:
                    position = int(parts[0])
                    value = int(parts[1])
                    events.append((position, value))
    return events

def add_tile_from_tape(board, position, value):
    """Add tile at specific position with specific value."""
    empty_positions = []
    for i in range(16):
        if ((board >> (4*i)) & 0xF) == 0:
            empty_positions.append(i)
    
    if not empty_positions:
        return board
    
    # Map tape position to actual empty cell
    actual_idx = empty_positions[position % len(empty_positions)]
    return board | (value << (4 * actual_idx))

def benchmark_expectimax(initial_board, num_iterations, search_depth):
    """Benchmark expectimax algorithm."""
    Game2048._init_tables()
    
    start_time = time.time()
    
    for _ in range(num_iterations):
        # Just compute the value, no actual moves
        value = max_value(initial_board, simple_eval, search_depth)
    
    end_time = time.time()
    elapsed = end_time - start_time
    
    return elapsed, value

def benchmark_game_with_tape(initial_board, tape_events, num_moves, search_depth):
    """Benchmark a deterministic game with tape."""
    Game2048._init_tables()
    
    start_time = time.time()
    
    board = initial_board
    tape_idx = 0
    score = 0
    moves_made = 0
    
    # Add initial tiles
    if tape_idx < len(tape_events):
        pos, val = tape_events[tape_idx]
        board = add_tile_from_tape(board, pos, val)
        tape_idx += 1
    
    if tape_idx < len(tape_events):
        pos, val = tape_events[tape_idx]
        board = add_tile_from_tape(board, pos, val)
        tape_idx += 1
    
    # Play game
    for move_num in range(num_moves):
        if Game2048.is_game_over(board):
            break
            
        move_name, _ = get_best_move(board, simple_eval, search_depth)
        if move_name is None:
            break
        
        # Execute move
        if move_name == 'up':
            board, move_score, moved = Game2048.move_up(board)
        elif move_name == 'down':
            board, move_score, moved = Game2048.move_down(board)
        elif move_name == 'left':
            board, move_score, moved = Game2048.move_left(board)
        elif move_name == 'right':
            board, move_score, moved = Game2048.move_right(board)
        
        if not moved:
            break
            
        score += move_score
        moves_made += 1
        
        # Add random tile from tape
        if tape_idx < len(tape_events):
            pos, val = tape_events[tape_idx]
            board = add_tile_from_tape(board, pos, val)
            tape_idx += 1
    
    end_time = time.time()
    elapsed = end_time - start_time
    
    return elapsed, board, score, moves_made

def main():
    print("Python Benchmark")
    print("=" * 50)
    
    # Test boards
    test_boards = [
        (0x0000000000000000, "empty"),
        (0x0000000000001234, "simple"),
        (0x1234567890ABCDEF, "complex"),
    ]
    
    # Benchmark expectimax
    print("\nExpectimax Benchmark (100 iterations):")
    for board, name in test_boards:
        for depth in [1, 2]:
            elapsed, value = benchmark_expectimax(board, 100, depth)
            print(f"  {name} board, depth {depth}: {elapsed:.3f}s (value: {value:.2f})")
    
    # Benchmark game with tape
    print("\nGame Benchmark with Tape:")
    tape_events = load_tape("test_tape.txt")
    
    for depth in [1, 2]:
        elapsed, final_board, score, moves = benchmark_game_with_tape(
            0x0000000000000000, tape_events, 50, depth
        )
        print(f"  Depth {depth}: {elapsed:.3f}s")
        print(f"    Final board: {final_board:016x}")
        print(f"    Score: {score}, Moves: {moves}")

if __name__ == "__main__":
    main()

File: benchmarks/benchmark_comparison.ml
(* Benchmark comparison for OCaml implementation *)

open Gp_2048_lib.Game_fast
open Gp_2048_lib.Gp_tree
open Gp_2048_lib.Expectimax_aligned
open Gp_2048_lib.Random_tape

(* Simple program that evaluates to NumEmptyCells + MaxTileValue *)
let simple_program = {
  nodes = [| Add; NumEmptyCells; MaxTileValue |];
  fitness = 0.0;
  games_played = 0;
  avg_score = 0.0;
  avg_max_tile = 0.0;
}


let benchmark_expectimax initial_board num_iterations search_depth =
  let start_time = Unix.gettimeofday () in
  
  let value = ref 0.0 in
  for _ = 1 to num_iterations do
    (* Just compute the value, no actual moves *)
    value := max_value initial_board simple_program search_depth
  done;
  
  let end_time = Unix.gettimeofday () in
  let elapsed = end_time -. start_time in
  (elapsed, !value)

let benchmark_game_with_tape initial_board tape num_moves search_depth =
  let start_time = Unix.gettimeofday () in
  
  let board = ref initial_board in
  let score = ref 0 in
  let moves_made = ref 0 in
  
  (* Add initial tiles *)
  let event1 = get_next_event tape in
  board := add_tile_with_event !board event1;
  let event2 = get_next_event tape in
  board := add_tile_with_event !board event2;
  
  (* Play game *)
  for _ = 1 to num_moves do
    if is_game_over !board then
      ()
    else
      match get_best_move !board simple_program search_depth with
      | None -> ()
      | Some dir ->
        let new_board = match dir with
          | `Up -> move_up !board
          | `Down -> move_down !board
          | `Left -> move_left !board
          | `Right -> move_right !board
        in
        
        if Int64.equal !board new_board then
          ()
        else begin
          let move_score = get_score_for_move !board dir in
          score := !score + move_score;
          board := new_board;
          incr moves_made;
          
          (* Add random tile from tape *)
          let event = get_next_event tape in
          board := add_tile_with_event !board event
        end
  done;
  
  let end_time = Unix.gettimeofday () in
  let elapsed = end_time -. start_time in
  (elapsed, !board, !score, !moves_made)

let main () =
  Printf.printf "OCaml Benchmark\n";
  Printf.printf "==================================================\n";
  
  (* Test boards *)
  let test_boards = [
    (0x0000000000000000L, "empty");
    (0x0000000000001234L, "simple");
    (0x1234567890ABCDEFL, "complex");
  ] in
  
  (* Benchmark expectimax *)
  Printf.printf "\nExpectimax Benchmark (100 iterations):\n";
  List.iter (fun (board, name) ->
    List.iter (fun depth ->
      let elapsed, value = benchmark_expectimax board 100 depth in
      Printf.printf "  %s board, depth %d: %.3fs (value: %.2f)\n" 
        name depth elapsed value
    ) [1; 2]
  ) test_boards;
  
  (* Benchmark game with tape *)
  Printf.printf "\nGame Benchmark with Tape:\n";
  let tape = load_random_tape "test_tape.txt" in
  
  List.iter (fun depth ->
    reset_tape tape;  (* Reset tape for each test *)
    let elapsed, final_board, score, moves = 
      benchmark_game_with_tape 0x0000000000000000L tape 50 depth in
    Printf.printf "  Depth %d: %.3fs\n" depth elapsed;
    Printf.printf "    Final board: %016Lx\n" final_board;
    Printf.printf "    Score: %d, Moves: %d\n" score moves
  ) [1; 2]

let () = main ()